### PR TITLE
Enhance data type

### PR DESCRIPTION
**Pull request by leehuwuj**
## Release Notes

- **New Features**
	- Enhanced file upload functionality with the introduction of the `DocumentFile` class, streamlining file metadata management and adding properties like `size` and `type`.
	- Updated file handling processes to improve clarity and consistency in metadata usage.
	- Improved document processing methods to directly utilize the new `DocumentFile` structure.

- **Bug Fixes**
	- Corrected file path generation in the code interpreter's results saving process.

- **Documentation**
	- Updated types and method signatures to reflect changes in file handling and metadata structures.

- **Refactor**
	- Simplified file saving logic and metadata processing across various components for better performance and clarity.

### CODE REVIEW

File: templates/types/streaming/nextjs/app/components/ui/chat/hooks/use-file.ts [hunk 1]
```diff
@@ -53,7 +46,8 @@ export function useFile() {
   const uploadContent = async (
     file: File,
     requestParams: any = {},
-  ): Promise<UploadedFileMeta> => {
+  ): Promise<DocumentFile> => {
+    // TODO: change API for Python and TS to return DocumentFile
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/hooks/use-file.ts:48]: remove once done

File: templates/types/streaming/nextjs/app/components/ui/document-preview.tsx [hunk 2]
```diff
@@ -23,9 +23,9 @@ export interface DocumentPreviewProps {
 }
 
 export function DocumentPreview(props: DocumentPreviewProps) {
-  const { filename, filesize, filetype, metadata } = props.file;
+  const { id, filename, filesize, filetype, metadata } = props.file;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/document-preview.tsx:26]: where is id used?

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx [hunk 3]
```diff
@@ -65,7 +66,7 @@ export default function ChatInput(
   };
 
   const handleUploadFile = async (file: File) => {
-    if (imageUrl || files.length > 0) {
+    if (imageUrl) {
       alert("You can only upload one file at a time.");
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx:70]:
```suggestion
      alert("You can only upload one image at a time.");
```

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx [hunk 4]
```diff
@@ -95,7 +96,7 @@ export default function ChatInput(
       )}
       {files.length > 0 && (
         <div className="flex gap-4 w-full overflow-auto py-2">
-          {files.map((file, index) => (
+          {files.map((file: DocumentFile) => (
             <DocumentPreview
               key={file.metadata?.id ?? `${file.filename}-${index}`}
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx]:
```suggestion
              key={file.id}
```

Comment (marcusschiesser) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx]: we also have `key={file.metadata?.id ?? `${file.filename}-${index}`}` somewhere else - also change it

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-files.tsx [hunk 5]
```diff
@@ -7,7 +7,7 @@ export function ChatFiles({ data }: { data: DocumentFileData }) {
     <div className="flex gap-2 items-center">
       {data.files.map((file, index) => (
         <DocumentPreview
-          key={file.metadata?.id ?? `${file.filename}-${index}`}
+          key={file.id ?? `${file.filename}-${index}`}
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/chat-message/chat-files.tsx:10]:
```suggestion
          key={file.id}
```

File: templates/types/streaming/fastapi/app/engine/utils/file_helper.py [hunk 6]
```diff
@@ -1,61 +1,61 @@
...
-from typing import Any, Dict, List, Optional
+from typing import List, Optional
 
-from pydantic import BaseModel, Field, computed_field
+from pydantic import BaseModel, Field, model_validator
 
 logger = logging.getLogger(__name__)
 
 
 class FileMetadata(BaseModel):
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/engine/utils/file_helper.py:12]: can this just be DocumentFile?

File: templates/types/streaming/fastapi/app/api/routers/upload.py [hunk 7]
```diff
@@ -14,22 +15,39 @@
 class FileUploadRequest(BaseModel):
     base64: str
     filename: str
+    filesize: Optional[int] = None
+    filetype: Optional[str] = None
     params: Any = None
 
 
+class DocumentFileResponse(BaseModel):
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/api/routers/upload.py:23]: difference to DocumentFile?

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx [hunk 8]
```diff
@@ -65,8 +66,8 @@ export default function ChatInput(
   };
 
   const handleUploadFile = async (file: File) => {
-    if (imageUrl || files.length > 0) {
-      alert("You can only upload one file at a time.");
+    if (imageUrl) {
+      alert("You can only upload one image at a time.");
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx:70]: @leehuwuj that's correct it's a bug, if the user uploads a CSV (but imageUrl is set), then can we still upload the CSV?

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts [hunk 9]
```diff
@@ -28,17 +28,17 @@ export const DOCUMENT_FILE_TYPES: DocumentFileType[] = [
 ];
 
 export type UploadedFileMeta = {
-  id: string;
   name: string; // The uploaded file name in the backend (including uuid and sanitized)
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/index.ts:31]: for what are we actually needing this? it's a bit confusing having a name and a filename

File: templates/types/streaming/nextjs/app/components/ui/chat/index.ts [hunk 10]
```diff
@@ -28,17 +28,17 @@ export const DOCUMENT_FILE_TYPES: DocumentFileType[] = [
...
 };
 
 export type DocumentFile = {
+  id: string;
   filename: string; // The original file name
-  filesize: number;
+  filesize: number; // The file size in bytes
   filetype: DocumentFileType;
-  metadata?: UploadedFileMeta; // undefined when the file is not uploaded yet
+  metadata: UploadedFileMeta;
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/index.ts:41]: how about inlining? 
```suggestion
   url?: string; // The URL of the uploaded file in the backend
  refs?: string[];
```

File: templates/types/streaming/nextjs/app/components/ui/chat/hooks/use-file.ts [hunk 11]
```diff
@@ -65,10 +56,12 @@ export function useFile() {
         ...requestParams,
         base64,
         filename: file.name,
+        filetype: docMineTypeMap[file.type],
+        filesize: file.size,
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/components/ui/chat/hooks/use-file.ts:60]: why does the backend need this?

File: templates/types/streaming/nextjs/app/api/chat/upload/route.ts [hunk 12]
```diff
@@ -14,16 +14,31 @@ export async function POST(request: NextRequest) {
...
     const index = await getDataSource(params);
-    return NextResponse.json(await uploadDocument(index, filename, base64));
+    const uploadedFileMeta = await uploadDocument(index, filename, base64);
+    const documentFile = {
+      id: uploadedFileMeta.id,
+      filename: filename, // Original filename
+      filesize,
+      filetype,
+      metadata: uploadedFileMeta,
+    };
```
Comment (marcusschiesser) [inline templates/types/streaming/nextjs/app/api/chat/upload/route.ts:40]:
```suggestion
    const documentFile = await uploadDocument(index, filename, base64);
```

File: templates/types/streaming/fastapi/app/engine/utils/file_helper.py (hunk 6 above)
Comment (marcusschiesser) [reply templates/types/streaming/fastapi/app/engine/utils/file_helper.py:12]: @leehuwuj bump

File: templates/types/streaming/fastapi/app/api/routers/upload.py [hunk 13]
```diff
@@ -14,22 +15,31 @@
...
         )
-        return file_meta.to_upload_response()
+
+        document_file = DocumentFile(
+            id=file_meta.id,
+            metadata=file_meta,
+            # Still return the original fields of the request to display in the chat UI.
+            filename=request.filename,
+            filesize=request.filesize,
+            filetype=request.filetype,
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/api/routers/upload.py:40]: should be easy to get file size on the server side - mime type can also be determined but might have an advantage to do this one the client side

File: templates/types/streaming/fastapi/app/api/routers/upload.py [hunk 14]
```diff
@@ -14,22 +15,31 @@
...
-def upload_file(request: FileUploadRequest) -> Dict[str, Any]:
+def upload_file(request: FileUploadRequest) -> DocumentFile:
     """
     To upload a private file from the chat UI.
-    Returns:
-        The metadata of the uploaded file.
     """
     try:
         logger.info(f"Processing file: {request.filename}")
         file_meta = PrivateFileService.process_file(
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/api/routers/upload.py:30]: return DocumentFile?

File: templates/components/llamaindex/typescript/streaming/annotations.ts [hunk 15]
```diff
@@ -6,14 +6,14 @@ export type DocumentFileType = "csv" | "pdf" | "txt" | "docx";
 export type UploadedFileMeta = {
   id: string;
```
Comment (marcusschiesser) [inline templates/components/llamaindex/typescript/streaming/annotations.ts:7]: why do we have here the id but not in Python?

File: templates/components/llamaindex/typescript/documents/upload.ts [hunk 16]
```diff
@@ -2,14 +2,15 @@ import { Document, LLamaCloudFileService, VectorStoreIndex } from "llamaindex";
...
+import { UploadedFileMeta } from "../streaming/annotations";
+import { parseFile, storeFile } from "./helper";
 import { runPipeline } from "./pipeline";
 
 export async function uploadDocument(
   index: VectorStoreIndex | LlamaCloudIndex | null,
   filename: string,
   raw: string,
-): Promise<FileMetadata> {
+): Promise<UploadedFileMeta> {
```
Comment (marcusschiesser) [inline templates/components/llamaindex/typescript/documents/upload.ts:13]: return DocumentFile?

File: templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (leehuwuj) [reply templates/types/streaming/nextjs/app/components/ui/chat/chat-input.tsx:70]: that's correct it's a bug, if the user uploads a CSV (but imageUrl is set), then can we still upload the CSV?
Nope, If the image is attached then you can not upload a new file - it'll throw the alert

File: templates/components/llamaindex/typescript/documents/helper.ts [hunk 17]
```diff
@@ -14,43 +15,38 @@ const MIME_TYPE_TO_EXT: Record<string, string> = {
...
+  name: string,
   fileBuffer: Buffer,
   mimeType: string,
 ) {
   const fileExt = MIME_TYPE_TO_EXT[mimeType];
   if (!fileExt) throw new Error(`Unsupported document type: ${mimeType}`);
 
   const fileId = crypto.randomUUID();
-  const newFilename = `${fileId}_${sanitizeFileName(filename)}`;
+  const newFilename = `${sanitizeFileName(name)}_${fileId}.${fileExt}`;
```
Comment (leehuwuj) [inline templates/components/llamaindex/typescript/documents/helper.ts:39]: The idea here is to move the file name above the ID for better display. I also removed duplicates in the original file name with the new file name. It's now always the name we stored on the server.

File: frontend/app/components/ui/chat/hooks/use-file.ts
Comment (marcusschiesser) [inline frontend/app/components/ui/chat/hooks/use-file.ts:1]: What's this?

Comment (leehuwuj) [reply frontend/app/components/ui/chat/hooks/use-file.ts:1]: Ah, that's Cursor bug adding file in other project again 😢

**Pull request merged**